Add basic lua tests

fidgetingbits opened the pull request:
This adds basic lua testing capabilities to fix https://github.com/cursorless-dev/cursorless/issues/2374, by using busted. Mostly followed along https://hiphish.github.io/blog/2024/01/29/testing-neovim-plugins-with-busted/

One thing I'm not sure about is that this will require busted and/or luarocks to be installed, so I don't know exactly where you want those to be listed as prerequisites and/or added to some developer setup script. I plan to use the flake.nix dev shell to get them for myself.

I plan to rewrite `buffer_get_selection()` entirely, but that doesn't need to be part of this PR. Having some existing tests for it lets me compare the new function behaves the same.

Something else I haven't done yet is adding a way to run it in CI, but that's partially because I'm not sure about the install part of luarocks, etc. I can do the CI part as part of this PR, or in a follow up.

Lastly I'll likely clean up the tests a little bit, as I can re-use some code in a few places, but figured I'd post this up anyway.

fidgetingbits committed:
initial lua test skeleton

fidgetingbits committed:
feat: test adding busted workflow

fidgetingbits commented:
Was just trying to run the lua tests I added via github workflow, and realized that the existing tests don't get far enough to run the ones I added:

```
 packages/cursorless-neovim build: > @cursorless/cursorless-neovim@1.0.0 populate-dist /home/runner/work/cursorless/cursorless/packages/cursorless-neovim
packages/cursorless-neovim build: > bash ./scripts/populate-dist.sh
packages/cursorless-neovim build: Populating dist directory...
packages/cursorless-neovim build: CURSORLESS_REPO_ROOT: /home/runner/work/cursorless/cursorless
packages/cursorless-neovim build: cp: cannot stat '/home/runner/work/cursorless/cursorless/cursorless.nvim/*': No such file or directory
packages/cursorless-neovim build:  ELIFECYCLE  Command failed with exit code 1.
packages/cursorless-neovim build: Failed
```

Will look into it, but if you have any ideas let me know.

pokey commented:
Yeah I broke the neovim CI a few days ago and haven't gotten around to fixing it yet sorry. I can have a look

fidgetingbits commented on cursorless.nvim/test/nvim-shim.sh:
```diff
@@ -0,0 +1,45 @@
+  ln -sf "${PWD}" ${link_path}
+fi
+
+NVIM=$(resolve_nvim_path)
```
If I use the flake.nix I used from https://github.com/saidelike/cursorless/pull/8 where I install bare nvim and force it to the beginning of PATH, I think I can skip the path resolution here. Will still need to invoke with `command nvim` though, but will simplify this a bunch.

saidelike commented on .gitignore:
```diff
@@ -46,3 +46,4 @@ packages/test-harness/testSubsetGrep.properties
 
 cursorless.nvim/node/cursorless-neovim
 cursorless.nvim/node/test-harness
+cursorless.nvim/test/xdg/
```
is that needed?

fidgetingbits replied:
Some folder is needed for the busted XDG environment variables and I was using that, but I can switch it to be a temp folder and remove this.

saidelike commented on docs/contributing/cursorless-in-neovim.md:
```diff
@@ -40,7 +40,13 @@ In order to test out your local version of the extension or to run unit tests lo
+
+### Running lua tests
+
+Their are separate cursorless and lua tests. You can run the lua tests by entering the `cursorless.nvim` folder and
```
maybe replace whole paragraph with:

There are separate specific neovim unit tests on top of the cursorless unit tests, which target the lua helper functions. In order to run these lua unit tests locally, enter the `cursorless.nvim` folder and run `busted --run unit`. These tests currently only work on Linux.

saidelike replied:
I know you marked this as resolved but I may want to clarify my thoughts:

- there are Cursorless tests for vscode that test the actions (`take`, `chuck` etc.)
- there are unit tests for Cursorless that tests specific typescipt functions (`toggle decorations`, `visible multiple regions`, etc.). These tests don't rely on vscode or neovim.
- there are Cursorless tests for neovim that test the actions (`take`, `chuck`, etc.). 99% of them are the same as the vscode ones except some new ones are needed in order to test all actions, due to no hats (and the fact that lots of Cursorless tests in vscode rely on hats)
- there are unit tests for lua in neovim. These tests are neovim specific and are testing lua functions. They need to be executed in neovim. There is no such tests for vscode because we assume the functions provided by [vscode APIs](https://code.visualstudio.com/api/references/vscode-api) have already been tested and work properly.

When you say `separate cursorless and lua tests` I understand that you mean `neovim cursorless and neovim lua tests`, right? Is that enough explanations? cc @pokey

fidgetingbits replied:
Yep, my feeling was that since the file is about neovim then them being contextual to neovim was implied, but happy to make it more explicit.

saidelike replied:
yeah, i think that's fine, but just want to make sure the distinction between `unit tests for Cursorless` and `unit tests for neovim` is clear. Because there is no `unit tests for vscode`, and we could technically be talking about `unit tests for Cursorless` above (instead of `unit tests for neovim`) since they are not vscode specific.

saidelike replied:
We could also describe all the different tests into a new documentation page somewhere in Cursorless docs, so we don't overload the neovim instructions above? cc @pokey

saidelike commented on cursorless.nvim/test/unit/cursorless_spec.lua:
```diff
@@ -0,0 +1,95 @@
+describe("cursorless tests", function()
+  local cursorless = require("cursorless.cursorless")
+
+  local get_selected_text = function()
```
Could the 2 local functions be in a separate file similarly to what is described in the "Local configuration" part of the blog post? So in `test/xdg/config/nvim`?

Once we have several .lua unit tests, these functions will be handy for all of these unit tests.

fidgetingbits replied:
Done. Now in `test/helpers.lua` and linked by the shim into a location that will load on neovim startup. Just exposed as global functions for now.

saidelike commented on cursorless.nvim/test/unit/cursorless_spec.lua:
```diff
@@ -0,0 +1,95 @@
+      function()
+        local lines = "hello world"
+        vim.api.nvim_buf_set_lines(0, 0, -1, false, vim.split(lines, "\n"))
+        cursorless.select_range(1, 4, 1, 2)
```
this uses another lua helper function so technically the unit test could fail because of this but not sure we can work around it?

fidgetingbits replied:
I guess we should add a test for each helper function and have those tests run first, so it's more apparent that a subsequent complex test that relies on the helper is possibly failing because of the earlier failure? Didn't address that yet

saidelike replied:
Happy if you want to create an issue for that and do it in a future PR or if you want to do it as part of this PR.

fidgetingbits replied:
I'll likely just do a follow up ya

fidgetingbits replied:
Filed an issue here: https://github.com/cursorless-dev/cursorless/issues/2548

saidelike commented:
That's amazing. Added a few comments in the code.

One drawback of having the unit tests inside `cursorless.nvim/` is that all these files will currently be copied when packaging the extension for production. So we could either move them elsewhere (if possible), or we will have to go back to copying only required files for production. cc @pokey

> One thing I'm not sure about is that this will require busted and/or luarocks to be installed, so I don't know exactly where you want those to be listed as prerequisites and/or added to some developer setup script. I plan to use the flake.nix dev shell to get them for myself.

What about adding requirements in `docs/contributing/cursorless-in-neovim.md` inside the `### Running lua tests` section?

> Something else I haven't done yet is adding a way to run it in CI, but that's partially because I'm not sure about the install part of luarocks, etc. I can do the CI part as part of this PR, or in a follow up.

I have fixed the neovim tests to work on CI in the nvim-talon branch. Can you rebase to the latest version so we confirm the lua unit tests work on CI?

> Lastly I'll likely clean up the tests a little bit, as I can re-use some code in a few places, but figured I'd post this up anyway.

What do you want to clean up?

saidelike commented on cursorless.nvim/test/unit/cursorless_spec.lua:
```diff
@@ -0,0 +1,95 @@
+  end
+
+  describe("window_get_visible_lines()", function()
+    before_each(function() end)
```
is that needed?

saidelike commented:
@fidgetingbits It might be worth rebasing it to nvim-talon too to confirm the tests pass

saidelike commented:
I did the rebase as a warning.

@fidgetingbits I would be curious if you know how to fix the failing tests now?

Commit:
fix: only install busted on Linux

fidgetingbits committed:
refactor: Address review feedback

fidgetingbits commented:
Addressed most of the feedback, but didn't look into why the CI test is failing yet.

fidgetingbits committed:
fix: change workflows to actually work

fidgetingbits commented:
I fixed the CI. The issue was that the busted action I chose to use turned out to be auto-running, so would fail before it got to the commands I had to manually run. I spent quite awhile trying to make it work, using `-C` to force it into `cursorless.nvim` folder and whatnot, but still could never get it to work, so just decided to go the manual lua/luarocks/busted install route. The downside is that it makes the test.yml a bit busier now.

There are at least a few options for cleaning it up a bit, so I figured I'd ask you before doing it:

1) Create a separate .yml file to break out these actions and then call it from test.yml
2) Create a new shell script to run the installations (which would mean manually calling apt get, etc) and call it from test.yml
3) Leave it as is

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,12 +55,27 @@ jobs:
+          cd cursorless.nvim
+          busted --run unit
+        if: runner.os == 'Linux' && matrix.app_version == 'stable'
+        id: neovimLuaTests
```
is this "id" useful?

fidgetingbits replied:
I don't think so since we don't reference it anywhere else. Will remove

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,6 +55,27 @@ jobs:
         if: runner.os == 'Linux' && matrix.app_version == 'stable'
         env:
           NEOVIM_PATH: ${{ steps.vim.outputs.executable }}
+      - name: Set up Lua
```
could remove that line

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,6 +55,27 @@ jobs:
+        with:
+          luaVersion: "luajit-2.1.0-beta3"
+        if: runner.os == 'Linux' && matrix.app_version == 'stable'
+      - name: Set up Luarocks
```
could remove that line

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,6 +55,27 @@ jobs:
+        with:
+          luaVersion: "5.1"
+        if: runner.os == 'Linux' && matrix.app_version == 'stable'
+      - name: Install Busted
```
could remove that line

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,6 +55,27 @@ jobs:
+      - name: Install Busted
+        run: |
+          luarocks install busted
+          luarocks install luafilesystem
```
could you merge the 2 "install" in a single line?

fidgetingbits replied:
Just notice this failed, so maybe can't be one line? Hadn't bothered tested it locally since it seemed sane, but will fix tomorrow.

saidelike replied:
https://github.com/luarocks/luarocks/issues/1359

saidelike replied:
that being said, if contents is into another test.yml, 2 lines is fine really.

saidelike replied:
or `luarocks install busted && luarocks install luafilesystem` lol

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,6 +55,27 @@ jobs:
+        uses: leafo/gh-actions-luarocks@v4
+        with:
+          luaVersion: "5.1"
+        if: runner.os == 'Linux' && matrix.app_version == 'stable'
```
@pokey are you aware of a way to avoid having to specify `if: runner.os == 'Linux' && matrix.app_version == 'stable'` for multiple steps as seen here and below?

fidgetingbits replied:
Seems possible by using a `using: composite` entry and then have a set of additional `steps:` entries, so will give it a try.

fidgetingbits replied:
Turns out you can only use composite if you break it out into a separate file, so open to other suggestions.

saidelike replied:
`uses: composite` with a separate `test_neovim_lua.yml` file might actually be nice?

fidgetingbits replied:
Ya I think it would be nicer/cleaner in a separate file personally. I can do it that way if you want

fidgetingbits replied:
Also is `using: composite` my bad

fidgetingbits replied:
This is done. I think the main downside is that it seems it has to go inside .github/actions/<something>/action.yml and you can't just name it whatever.yml. But otherwise is cleaner imo.

saidelike commented on .github/workflows/test.yml:
```diff
@@ -55,6 +55,27 @@ jobs:
+      - name: Run cursorless.nvim busted lua tests
+        run: |
+          cd cursorless.nvim
+          busted --run unit
```
could these 2 commands be executed in a single line? `run: cd cursorless.nvim && busted --run unit`?

saidelike commented:
> I fixed the CI. The issue was that the busted action I chose to use turned out to be auto-running, so would fail before it got to the commands I had to manually run. I spent quite awhile trying to make it work, using `-C` to force it into `cursorless.nvim` folder and whatnot, but still could never get it to work, so just decided to go the manual lua/luarocks/busted install route. The downside is that it makes the test.yml a bit busier now.

Nice one :)

> 
> There are at least a few options for cleaning it up a bit, so I figured I'd ask you before doing it:
> 
>     1. Create a separate .yml file to break out these actions and then call it from test.yml
> 
>     2. Create a new shell script to run the installations (which would mean manually calling apt get, etc) and call it from test.yml
> 
>     3. Leave it as is

I think leaving it as-is is good.

1. test.yml is not that big atm imho
2. that would make it more annoying as we would not rely on power of test.yml which simplify the amount of code we need to write

I am happy with the PR and it is close to be merged imho, only a few minor feedbacks and we are ready to go!

pokey commented on docs/contributing/cursorless-in-neovim.md:
```diff
@@ -40,7 +40,13 @@ debug mode. To do so you need to run the `workbench.action.debug.selectandstart`
+### Running lua tests
+
+Their are separate cursorless and lua tests. You can run the lua tests by entering the `cursorless.nvim` folder and
+running: `busted --run unit`. These tests currently only work on Linux.
```
Should we add a vscode launch config or task for this?

fidgetingbits replied:
Sure, will do as part of this PR

fidgetingbits committed:
fix: Address review feedback

saidelike commented:
@fidgetingbits Just to let you know I have fixed Cursorless tests in neovim again on CI so merging nvim-talon into this PR will fix the tests here too.

saidelike commented on .vscode/tasks.json:
```diff
@@ -268,6 +268,14 @@
       }
     },
+    {
+      "label": "Neovim: Run cursorless.nvim lua tests",
```
"label": "Neovim: Test Lua"

saidelike commented:
@fidgetingbits Do you mind documenting the architecture of the lua tests in https://github.com/saidelike/cursorless/blob/nvim-talon/docs/contributing/architecture/tests.md too?

fidgetingbits committed:
Merge remote-tracking branch 'saidelike/nvim-talon' into lua-tests

fidgetingbits committed:
fix: Address task.json feedback

fidgetingbits committed:
docs: Add neovim lua test architecture section

fidgetingbits committed:
refactor: switch to standalone neovim lua test action

fidgetingbits commented:
> @fidgetingbits Do you mind documenting the architecture of the lua tests in https://github.com/saidelike/cursorless/blob/nvim-talon/docs/contributing/architecture/tests.md too?

I took a stab at it.

fidgetingbits commented on docs/contributing/architecture/tests.md:
````diff
@@ -31,7 +31,8 @@ packages/test-harness/src/config/init.lua
   -> TestHarnessRun() -> run() -> runAllTests() -> Mocha + packages/cursorless-neovim-e2e/src/suite/recorded.neovim.test.ts
 ```
 
-4. **Lua unit tests for neovim**: tests lua functions. They need to be executed in neovim. There is no such test for VSCode because we assume the functions provided by [VSCode APIs](https://code.visualstudio.com/api/references/vscode-api) have already been tested by Microsoft and work properly
````
Just to note I removed this part about VSCode APIs because I think it's irrelevant. It almost implies we are testing neovim builtin APIs or something. The lua tests for cursorless.nvim are more analogous to the cursorless (non-action) unit tests imo.

saidelike replied:
Yes, it is interesting. 

The reason I say the "lua helper functions" in neovim are similar to the "vscode APIs" in vscode is because we have to basically emulate the vscode APIs by combining lots of lua API calls. 

This is because Cursorless was built for VSCode in the first place so is very dependent on VSCode APIs architecture for now. It will probably change at some point and there may be "typescript unit tests for vscode" later (which would test combining lots of typescript calls, but specifically testing in the case of vscode, though it would indeed also be a way to test the lua calls in the case of neovim, and in that sense I agree they will be closer to the cursorless unit tests).

So ya, I think long term the lua tests for Cursorless neovim will be similar to the typescript unit tests for vscode. But we are not there yet I think.

saidelike replied:
Curious what is your view on this @pokey?

fidgetingbits replied:
Thanks for the explanation. fwiw even with that explanation, imo, including what I deleted still doesn't make sense. If you want to highlight that neovim side needs custom lua functions that mimic vscode api and those need to be tested and so is why these tests are for, then I'd change the wording to just say that instead. saying we don't need to test vscode's api and linking to those api is still quite confusing (to me).

saidelike replied:
Yes, I think it is better with old text removed. But I just thought I would explain. So fine with left as is.

fidgetingbits committed:
fix: Remove unneeded luaVersion entry for luarocks

fidgetingbits commented:
Tests are passing and is easier to read now (imo), so hopefully mostly good to go

saidelike commented on .vscode/tasks.json:
```diff
@@ -287,6 +287,14 @@
       // NOTE: We don't have a way on Windows atm due to command with argument inside Run() not working
       // so we need to show logs outside of vscode (see #2454)
     },
+    {
```
Can you run this within vscode GUI? I don't see it on mine (I know I won't be able to test it on Windows though). Or maybe you need to define a launch config for that?

fidgetingbits replied:
Ya I choose run task from the command pallette and then it shows up.

saidelike replied:
Ah ok ya would be better to expose it as a launch config. I can try to do it for you to test if you prefer too?

fidgetingbits replied:
Looks like launch.json is specifically for debugging, so you want to be able to debug while the tests are running ya? I guess it's just a matter of attaching to node and executed busted with the debug environment variable set.

saidelike replied:
Right, yes, `launch.json` is for debugging. And yes that's what we want. Something similar to the `Neovim: Run` task I guess?

How does `busted` execute `nvim`? Is it only one execution of `nvim` for all tests (that are part of all the _spec.lua files) or does it spawn one `nvim` for each test? I guess we can only attach to it if only one execution of `nvim`? cc @pokey

fidgetingbits replied:
My understanding is that one instance of lua (so nvim) is run for each `*_spec.lua` file. So we could choose to just put everything in one, as busted does provide some nice mechanisms for organizing tests inside the spec file. Still might get a bit out of hand eventually though..

fidgetingbits replied:
https://github.com/cursorless-dev/cursorless/issues/2570

saidelike commented on cursorless.nvim/test/nvim-shim.sh:
```diff
@@ -0,0 +1,28 @@
+ln -sf "${PWD}/test/helpers.lua" "${plugin_folder}/helpers.lua"
+
+# shellcheck disable=SC2068
+command nvim --cmd 'set loadplugins' -l $@
```
does using `-V1` allow having better output than this?

```
Run cd cursorless.nvim
:noh
:'<,'>noh
:'<,'>noh
+++++
5 successes / 0 failures / 0 errors / 0 pending : 0.002549 seconds
1, 3, 1, 5, true
```

it woud be nice if we could at least have the name of each test printed so we know what is failing once it happens.

See https://neovim.io/doc/user/starting.html#-l

fidgetingbits replied:
I'll look into it. I actually don't get why the output from each test that is there gets printed as is even. Would be nice if the noh stuff could be silenced. 

I guess the most important thing is more what it looks like when one fails so it's easily identifiable. But ya may as well print out each passing test too. I kind of expect busted to do something like that instead of having to print and show with -V1 ourselves, but will see.

saidelike replied:
Yes, I guess it would be nice to know what the output is when a test fails if you can share? I can't test myself.

fidgetingbits replied:
I fixed all the output now, so it's not got the extra cruft. I didn't bother using a custom output handler, I just use the [TAP](https://testanything.org/) one, which is a well defined standard. I also no longer defer output, so it prints each test as ok/not ok as it runs them.

A failed output looks like this:

```
ok 1 - cursorless.nvim tests window_get_visible_lines() Can read one visible line
ok 2 - cursorless.nvim tests window_get_visible_lines() Can read all lines visible on the window
ok 3 - cursorless.nvim tests select_range() Selects the specified range
ok 4 - cursorless.nvim tests buffer_get_selection() Can get the forward selection in a format expected by cursorless
not ok 5 - cursorless.nvim tests buffer_get_selection() Can get the backward selection in a format expected by cursorless
# ./test/unit/cursorless_spec.lua @ 57
# Failure message: ./test/unit/cursorless_spec.lua:63: assertion failed!
1..5
E5113: Error while calling lua chunk: [NULL]
```

The only thing I don't like (minor nit) about the TAP output is the `1..5` at the end, which indicates it ran tests 1 through 5, as it's a bit confusing at first. I thought it was more vim spam lol.

saidelike replied:
I think it is fine like this <3

Commit:
docs: review busted tests documentation

saidelike commented:
Great stuff. I've changed a bit the tests documentation based on my understanding, hopefully you agree with my changes https://github.com/saidelike/cursorless/pull/7/commits/3bef6603186b8e1faea7b8935f06cfb0cc351530 feel free to review/tweak otherwise. cc @fidgetingbits

fidgetingbits committed:
refactor: print each test name as they run

fidgetingbits commented:
Updated to address some of the feedback. Still no launch.json, but tests output are improved:

```bash
❯ busted --run unit
ok 1 - cursorless.nvim tests window_get_visible_lines() Can read one visible line
ok 2 - cursorless.nvim tests window_get_visible_lines() Can read all lines visible on the window
ok 3 - cursorless.nvim tests select_range() Selects the specified range
ok 4 - cursorless.nvim tests buffer_get_selection() Can get the forward selection in a format expected by cursorless
ok 5 - cursorless.nvim tests buffer_get_selection() Can get the backward selection in a format expected by cursorless
1..5
```

I also added `.busted` as a lua file in `settings.json` as it was defaulting to javascript for whatever reason.

I didn't review the documentation changes yet.

Also fwiw regarding being unable to test, I suspect you could run busted from wsl on windows no problem, assuming you actually want to try it. I suppose we should file an issue so that we don't rely on the XDG environment to run busted under in the future, and that way those tests can be run on windows, mac, etc.

fidgetingbits commented:
> ```shell
> ❯ busted --run unit
> ok 1 - cursorless.nvim tests window_get_visible_lines() Can read one visible line
> ok 2 - cursorless.nvim tests window_get_visible_lines() Can read all lines visible on the window
> ok 3 - cursorless.nvim tests select_range() Selects the specified range
> ok 4 - cursorless.nvim tests buffer_get_selection() Can get the forward selection in a format expected by cursorless
> ok 5 - cursorless.nvim tests buffer_get_selection() Can get the backward selection in a format expected by cursorless
> 1..5
> ```

Now that I know how this output looks, I'll tweak it to be even more concise/clean, but didn't have time today.

saidelike commented:
> I didn't review the documentation changes yet.

You might want to merge main into nvim-talon before doing so as Pokey has modified its contents a bit. See https://github.com/cursorless-dev/cursorless/pull/2256/commits/ee52d25598efa479fbf1689d9aa98aba86c4f738

saidelike commented:
> > ```shell
> > ❯ busted --run unit
> > ok 1 - cursorless.nvim tests window_get_visible_lines() Can read one visible line
> > ok 2 - cursorless.nvim tests window_get_visible_lines() Can read all lines visible on the window
> > ok 3 - cursorless.nvim tests select_range() Selects the specified range
> > ok 4 - cursorless.nvim tests buffer_get_selection() Can get the forward selection in a format expected by cursorless
> > ok 5 - cursorless.nvim tests buffer_get_selection() Can get the backward selection in a format expected by cursorless
> > 1..5
> > ```
> 
> Now that I know how this output looks, I'll tweak it to be even more concise/clean, but didn't have time today.

Note sure how you can improve it but I like already way more how it is now, nice one :)

Is it possible to simulate a failure to see how it renders?

fidgetingbits commented:
> Is it possible to simulate a failure to see how it renders?

I had put it in a response here https://github.com/saidelike/cursorless/pull/7#discussion_r1689771749 already

Commit:
refactor: rename task

Commit:
refactor: better output for lua tests

Commit:
Merge branch 'nvim-talon' into lua-tests
